src/main/java/com/team701/buddymatcher/socket/ChatEventHandler.java

```diff
@@ -34,37 +48,115 @@ public ChatEventHandler(SocketIOServer server, MessageRepository messageReposito
         this.namespace.addConnectListener(onConnected());
         this.namespace.addDisconnectListener(onDisconnected());
         this.namespace.addEventListener(MESSAGE_EVENT_KEY, MessageObject.class, onMessageReceived());
+        this.namespace.addEventListener(READ_EVENT_KEY, ReadObject.class, onReadReceived());
+
+        this.messageRepository = messageRepository;
     }
 
+    /**
+     * onMessageReceived listener that sends message to the receiver when it receives a message event
+     * @return
+     */
     private DataListener<MessageObject> onMessageReceived() {
         return (client, data, ackSender) -> {
-            log.debug("Client[{}] - Received chat message '{}'", client.getSessionId().toString(), data);
+            log.debug("Client[{}] - Received message '{}'", client.getSessionId().toString(), data);
 
-            SocketIOClient receiver = namespace.getClient(userConnections.get(data.receiverId));
-            if (receiver != null) receiver.sendEvent(MESSAGE_EVENT_KEY, data);
+            SocketIOClient receiver = namespace.getClient(userIdConnections.get(data.getReceiverId()));
+            if (receiver != null) sendMessage(receiver, data);
             putMessageInHistory(data);
         };
     }
 
+    /**
+     * onReadReceived listener that sends read notice to the buddy when it receives a read event
+     * @return
+     */
+    private DataListener<ReadObject> onReadReceived() {
+        return (client, data, ackSender) -> {
+            Long userId = connectionUserIds.get(client.getSessionId());
+            log.debug("Client[{}] User[{}] - Read messages", client.getSessionId().toString(), userId);
+
+            messageRepository.updateUnreadMessagesForAUser(userId, data.getBuddyId());
+
+            SocketIOClient receiver = namespace.getClient(userIdConnections.get(data.getBuddyId()));
+            if (receiver != null) sendRead(receiver, new ReadObject().setBuddyId(userId));
+        };
+    }
+
+    /**
+     * onConnected listener that stores current user connected and reports online users to all users
+     * @return
+     */
     private ConnectListener onConnected() {
         return client -> {
             HandshakeData handshakeData = client.getHandshakeData();
-            log.debug("Client[{}] - Connected to chat handler through '{}'", client.getSessionId().toString(), handshakeData.getUrl());
+            String userId = handshakeData.getSingleUrlParam(USER_ID_KEY);
+            UUID clientId = client.getSessionId();
+
+            // Check if userId is not a number, report invalid userId
+            if (!userId.matches("\\d*")) {
```

Comment:
I'm not sure if this regex will catch all valid `Long`s. You could use a `try`/`catch` when you parse it to a Long instead.
Ref: https://stackoverflow.com/a/2563675